### PR TITLE
fix(server): re-add proxy memo

### LINKED ISSUE #5808: bug: Infinite render loop when referencing utils as dependency
### Provide environment information

  System:
    OS: macOS 14.5
    CPU: (10) arm64 Apple M2 Pro
    Memory: 91.06 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.19.0 - ~/Library/Caches/fnm_multishells/3059_1718376229215/bin/node
    npm: 10.2.3 - ~/Library/Caches/fnm_multishells/3059_1718376229215/bin/npm
  Browsers:
    Brave Browser: 121.1.62.165
    Chrome: 126.0.6478.62
    Edge: 126.0.2592.61
    Safari: 17.5
  npmPackages:
    @tanstack/react-query: ^5.39.0 => 5.45.1
    @trpc/client: next => 11.0.0-rc.413+87f0cbc94
    @trpc/next: next => 11.0.0-rc.413+87f0cbc94
    @trpc/react-query: next => 11.0.0-rc.413+87f0cbc94
    @trpc/server: next => 11.0.0-rc.413+87f0cbc94
    next: ^14.2.1 => 14.2.4
    react: ^18.3.0 => 18.3.1
    typescript: ^5.4.2 => 5.4.5

### Describe the bug

When hook dependencies contain the result of the `trpc.useUtils` call, react will run into an infinite loop.

Expectation:

The returned objects from trpc.useUtils are stable and can be used as dependencies.

### Link to reproduction

https://github.com/Csszabi98/trpc-infinite-loop-bug

### To reproduce

```typescript
const hello = api.post.hello.useQuery({ text: "from tRPC" });
const utils = trpc.useUtils()

useEffect(() => {
   utils.post.hello.invalidate()
}, [utils.post.hello])

```

### Additional information

_No response_

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/5808">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/5808/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by KATT**
Closes #5808

## 🎯 Changes

- Re-adds proxy memorization with more defensive handling - a mutation of the arguments caused an issue
- Adds `readonly` prefix on a bunch of places to prevent mutation of arguments' arrays

### CODE REVIEW

File: packages/tests/server/react/rsc-prefetch.test.tsx [hunk 1]
```diff
@@ -100,13 +101,22 @@ test('rsc prefetch helpers', async () => {
...
+  const utils1 = render(
+    <App>
+      <Parent />
+    </App>,
+  );
+  await waitFor(() => {
+    expect(utils1.container).toHaveTextContent(`__result`);
+  });
+
+  const utils2 = render(
```
Comment (juliusmarminge) [inline packages/tests/server/react/rsc-prefetch.test.tsx:113]: logs out this from the log I added above:

image

Comment (KATT) [reply packages/tests/server/react/rsc-prefetch.test.tsx:113]: hmmm, this fails even if I revert the proxy caching here

File: packages/react-query/src/rsc.tsx [hunk 2]
```diff
@@ -100,7 +100,9 @@ export function createHydrationHelpers<TRouter extends AnyRouter>(
 
   const wrappedProxy = createRecursiveProxy<
     DecorateRouterRecord<RootTypes, TRouter['_def']['record']>
-  >(({ path, args }) => {
+  >(async (opts) => {
+    const path = [...opts.path];
+    const args = [...opts.args];
```
Comment (KATT) [inline packages/react-query/src/rsc.tsx:105]: This is the actual fix

File: packages/server/src/unstable-core-do-not-import/createProxy.ts [hunk 3]
```diff
@@ -1,26 +1,34 @@
...
+  args: readonly unknown[];
 }
 type ProxyCallback = (opts: ProxyCallbackOptions) => unknown;
 
 const noop = () => {
   // noop
 };
 
+const freezeIfAvailable = (obj: object) => {
+  if (Object.freeze) {
```
Comment (juliusmarminge) [inline packages/server/src/unstable-core-do-not-import/createProxy.ts:12]: typeof checks are usually better for checking if globals exist but in this case Object is always defined so this should work?

File: packages/tests/server/react/rsc-prefetch.test.tsx [hunk 4]
```diff
@@ -78,36 +87,48 @@ const ctx = konn()
...
   );
   await waitFor(() => {
-    expect(utils.container).toHaveTextContent(`__result`);
+    expect(utils1.container).toHaveTextContent('__result1');
+    expect(utils1.container).toHaveTextContent('__result2');
   });
+  expect(ctx.postByIdInvokations).toHaveBeenCalledTimes(2);
 
   // Should not have fetched from CC but taken promise from server client
   expect(fetchSpy).toHaveBeenCalledTimes(0);
```
Comment (juliusmarminge) [inline packages/tests/server/react/rsc-prefetch.test.tsx:134]: Can you verify this is the case "in the real world" by running the example and checking that it doesn't fetch?

I should add an e2e test 😅

**Pull request merged**